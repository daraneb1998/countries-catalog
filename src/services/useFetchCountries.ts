import { ref } from "vue";
import type { CountryInfoType } from "@/types";

export const useFetchCountries = () => {
  const error = ref<string>("");
  const isFetching = ref(false);

  const fetchCountryData = async () => {
    isFetching.value = true;
    try {
      const response = await fetch("https://restcountries.com/v3.1/all");
      if (!response.ok) {
        throw new Error("Failed to fetch country data!");
      }
      const responseData = (await response.json()) as Array<CountryInfoType>;
      return responseData;
    } catch (err) {
      error.value = (err as Error).message;
      return [];
    } finally {
      isFetching.value = false;
    }
  };

  return {
    error,
    isFetching,
    fetchCountryData,
  };
};
