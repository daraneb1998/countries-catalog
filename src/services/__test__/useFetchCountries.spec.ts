import { describe, it, expect, beforeEach, vi } from "vitest";

import { useFetchCountries } from "../useFetchCountries";
import { mockedCountry } from "@/assets/data";

describe("useFetchCountries", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("fetches data successfully", async () => {
    global.fetch = vi.fn().mockResolvedValue({
      ok: true,
      json: vi.fn().mockResolvedValue(mockedCountry),
    });
    const { isFetching, error, fetchCountryData } = useFetchCountries();

    const data = await fetchCountryData();
    expect(data).toEqual(mockedCountry);
    expect(isFetching.value).toBe(false);
    expect(error.value).toEqual("");
  });

  it("handles fetch failure", async () => {
    global.fetch = vi.fn().mockResolvedValue({
      ok: false,
    });
    const { isFetching, error, fetchCountryData } = useFetchCountries();

    const data = await fetchCountryData();

    expect(data).toEqual([]);
    expect(isFetching.value).toBe(false);
    expect(error.value).toBe("Failed to fetch country data!");
  });

  it("handles network error", async () => {
    global.fetch = vi.fn().mockRejectedValue(new Error("Network error"));
    const { isFetching, error, fetchCountryData } = useFetchCountries();

    const data = await fetchCountryData();

    expect(data).toEqual([]);
    expect(isFetching.value).toBe(false);
    expect(error.value).toBe("Network error");
  });
});
