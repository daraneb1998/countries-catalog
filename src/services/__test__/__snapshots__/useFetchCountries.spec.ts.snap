// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CountryList > renders properly 1`] = `""`;

exports[`useFetchCountries > Initialize state 1`] = `
RefImpl {
  "__v_isRef": true,
  "__v_isShallow": false,
  "_rawValue": "",
  "_value": "",
  "dep": undefined,
}
`;
